lib/Sema/CSApply.cpp

```diff
@@ -7044,27 +7029,28 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
       }
     }
 
-    /// Whether the given effect should be propagated to a closure expression.
-    auto shouldApplyEffect = [&](EffectKind kind) -> bool {
+    /// Whether the given effect is polymorphic at this location.
+    auto isEffectPolymorphic = [&](EffectKind kind) -> bool {
       auto last = locator.last();
       if (!(last && last->is<LocatorPathElt::ApplyArgToParam>()))
-        return true;
+        return false;
 
-      // The effect should not be applied if the closure is an argument
-      // to a function where that effect is polymorphic.
       if (auto *call = getAsExpr<ApplyExpr>(locator.getAnchor())) {
         if (auto *declRef = dyn_cast<DeclRefExpr>(call->getFn())) {
           if (auto *fn = dyn_cast<AbstractFunctionDecl>(declRef->getDecl()))
-            return !fn->hasPolymorphicEffect(kind);
+            return fn->hasPolymorphicEffect(kind);
```

Comment:
IHMO naming of this method could be improved it's clear what polymorphism has to do with this and there is no comment at the declaration site. @hborla told me that this is about interaction between reasync/rethrows and async/throws attributes.